automates/model_analysis/identification_algorithms.py

```diff
@@ -0,0 +1,251 @@
+        tree.children.append(nxt.tree)
+        tree.call.line = 3
+        tree.call.id_check = nxt.tree.call.id_check
+        tree.call.w = w
+        tree.call.an_xbar = an_xbar
+        return gm.ResultsInternal(p=nxt.p, tree=tree)
+
+    # Line 4
+    g_remove_x = g.subgraph(list(set(v) - set(x)))
+    s = gm.c_components(g_remove_x, topo)
+    if len(s) > 1:
+        tree.call.line = 4
+        product_list = []
+        id_check_list = []
+        for s_element in s:
+            nxt = compute_ID(s_element, gm.ts(set(v) - set(s_element), topo), p, g, g_obs, v, topo, gm.TreeNode())
+            product_list.append(nxt.p)
+            id_check_list.append(nxt.tree.call.id_check)
+            tree.children.append(nxt.tree)
+        tree.call.id_check = all(id_check_list)
+        return gm.ResultsInternal(p=gm.Probability(sumset=gm.ts(list(set(v) - set(y) - set(x)), topo), product=True, \
+                                                   children=product_list), tree=tree)
+    else:
+        s_single = s[0]
+
+        # Line 5
+        cc = gm.c_components(g, topo)
+        if cc[0] == v:
+            tree.call.s = cc[0]
+            tree.call.line = 5
+            tree.call.id_check = False
+            tree.root = deepcopy(p)
+            return gm.ResultsInternal(p=p, tree=tree)
+
+        # Line 6
+        if s_single in cc:
+            tree.call.line = 6
+            tree.call.s = s_single
+            s_single_length = len(s_single)
+            product_list = []
+            p_prod = gm.Probability()
+            for node in s_single:
+                node_topo_index = v.index(node)
+                cond_set = v[0:node_topo_index]
+                if p.product:
+                    p_prod = gm.parse_joint(p, node, cond_set, v, topo)
+                else:
+                    p_prod = deepcopy(p)
+                    p_prod.var = [node]
+                    p_prod.cond = cond_set
+                product_list.append(p_prod)
+            product_list.reverse()
+            if s_single_length > 1:
+                prob_new = gm.Probability(sumset=gm.ts(set(s_single) - set(y), topo), product=True, children= \
+                    product_list)
+                tree.root = prob_new
+                tree.call.id_check = True
+                return gm.ResultsInternal(p=prob_new, tree=tree)
+            if p_prod.product or p_prod.fraction:
+                p_prod.sumset = gm.ts(list(set(p_prod.sumset) | (set(s_single) - set(y))), topo)
+            else:
+                p_prod.var = gm.ts(set(p_prod.var) - (set(p_prod.sumset) | (set(s_single) - set(y))), topo)
+            tree.root = p_prod
+            tree.call.id_check = True
+            return gm.ResultsInternal(p=p_prod, tree=tree)
+
+        # Line 7
+        tree.call.s = s_single
+        s_prime = None
+        for component in cc:
+            if set(s_single).issubset(set(component)):
+                s_prime = component
+                break
```

Comment:
@aswinchester for safety I would add a check after this loop to see if `s_prime` was never assigned. That may save some heartache during future debugging if an example goes wrong. The following would work:
```python
if s_prime is None:
  raise RuntimeError("s_prime component not identified in cc")
```

Reply:
This has been completed